leader/src/jerigon.rs

```diff
@@ -1,61 +1,64 @@
-use std::{
-    fs::{create_dir_all, File},
-    io::Write,
-    path::PathBuf,
-};
+use std::path::PathBuf;
 
 use alloy::providers::RootProvider;
 use anyhow::Result;
 use common::block_interval::BlockInterval;
+use common::fs::generate_block_proof_file_name;
 use paladin::runtime::Runtime;
 use proof_gen::proof_types::GeneratedBlockProof;
+use tracing::{error, info};
+
+#[derive(Debug, Default)]
+pub(crate) struct JerigonParams {
+    pub checkpoint_block_number: u64,
+    pub previous_proof: Option<GeneratedBlockProof>,
+    pub proof_output_dir: Option<PathBuf>,
+    pub save_inputs_on_error: bool,
+    pub keep_intermediate_proofs: bool,
+}
 
 /// The main function for the jerigon mode.
 pub(crate) async fn jerigon_main(
     runtime: Runtime,
     rpc_url: &str,
     block_interval: BlockInterval,
-    checkpoint_block_number: u64,
-    previous_proof: Option<GeneratedBlockProof>,
-    proof_output_dir_opt: Option<PathBuf>,
-    save_inputs_on_error: bool,
+    mut params: JerigonParams,
 ) -> Result<()> {
     let prover_input = rpc::prover_input(
         RootProvider::new_http(rpc_url.parse()?),
         block_interval,
-        checkpoint_block_number.into(),
+        params.checkpoint_block_number.into(),
     )
     .await?;
 
-    let block_proofs = prover_input
-        .prove(&runtime, previous_proof, save_inputs_on_error)
+    // If `keep_intermediate_proofs` is not set we only keep the last block
+    // proof from the interval. It contains all the necessary information to
+    // verify the whole sequence.
+    let proved_blocks = prover_input
+        .prove(
+            &runtime,
+            params.previous_proof.take(),
+            params.save_inputs_on_error,
+            params.proof_output_dir.clone(),
+        )
         .await?;
     runtime.close().await?;
 
-    for block_proof in block_proofs {
-        let block_proof_str = serde_json::to_vec(&block_proof)?;
-        write_proof(
-            block_proof_str,
-            proof_output_dir_opt.clone().map(|mut path| {
-                path.push(format!("b{}.zkproof", block_proof.b_height));
-                path
-            }),
-        )?;
-    }
-    Ok(())
-}
-
-fn write_proof(proof: Vec<u8>, proof_output_dir_opt: Option<PathBuf>) -> Result<()> {
-    match proof_output_dir_opt {
-        Some(p) => {
-            if let Some(parent) = p.parent() {
-                create_dir_all(parent)?;
-            }
+    info!("Proved blocks: {proved_blocks:?}");
 
-            let mut f = File::create(p)?;
-            f.write_all(&proof)?;
+    if !params.keep_intermediate_proofs {
+        if let Some(proof_output_dir) = params.proof_output_dir.as_ref() {
```

Comment:
minor: maybe a `warn!` message if `params.proof_output_dir` is `None`, given that we want to remove non-existing intermediate proofs.